### PR TITLE
Flink: deprecate ReaderFunction with a new Converter interface to simplify user experience

### PR DESCRIPTION
**Pull request by stevenzwu**

### CODE REVIEW

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowDataConverter.java [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  class IdentityConverter implements RowDataConverter<RowData> {
+    @Override
+    public RowData apply(RowData from) {
+      return from;
+    }
+
+    @Override
+    public TypeInformation getProducedType() {
+      return TypeInformation.of(RowData.class);
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowDataConverter.java:47]: Don't we need the `RowType` here?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterReaderFunction.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+    this.filters = filters;
+    this.limit = limit;
+  }
+
+  @Override
+  protected DataIterator<T> createDataIterator(IcebergSourceSplit split) {
+    RowDataFileScanTaskReader rowDataReader =
+        new RowDataFileScanTaskReader(tableSchema, readSchema, nameMapping, caseSensitive, filters);
+    return new LimitableDataIterator<>(
+        new ConverterFileScanTaskReader<>(rowDataReader, converter),
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterReaderFunction.java:75]: Would it worth to add the converter to the reader instead of adding a new wrapper? Getting a bit hard to follow this many embedded readers

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowDataConverter.java (hunk 1 above)
Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/RowDataConverter.java:47]: you are right. we need to do sth like this.
```
      TypeInformation<RowData> typeInfo =
          FlinkCompatibilityUtil.toTypeInfo(FlinkSchemaUtil.convert(readSchema));
```

I will actually remove the whole `IdentityConverter`. earlier I was thinking about create a default `IdentityConverter` if the converter is null. but that wasn't needed anymore.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterReaderFunction.java (hunk 2 above)
Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterReaderFunction.java:75]: I have thought it too. To execute the converter in this method, we would need to override/extend the `DataIterator`. That is also non trivial, as `DataIterator` is not a pure simple `CloseableIterator`. Running the converter inside the `ConverterFileScanTaskReader` is a little simpler.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterFileScanTaskReader.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+import org.apache.flink.annotation.Internal;
+import org.apache.iceberg.FileScanTask;
+import org.apache.iceberg.encryption.InputFilesDecryptor;
+import org.apache.iceberg.flink.source.FileScanTaskReader;
+import org.apache.iceberg.flink.source.RowDataFileScanTaskReader;
+import org.apache.iceberg.io.CloseableIterator;
+
+@Internal
+public class ConverterFileScanTaskReader<T> implements FileScanTaskReader<T> {
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterFileScanTaskReader.java:29]: Maybe add this as an internal class to the `ConverterReaderFunction`? Both classes are small, and not used elsewhere, so this could help us keeping the source file number lower.

Not a strong opinion, just an idea

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java [hunk 4]
```diff
@@ -69,12 +70,13 @@ public class TestIcebergSourceBoundedGenericRecord {
...
+  @Parameters(name = "format={0}, parallelism = {1}, useConverter = {2}")
   public static Object[][] parameters() {
     return new Object[][] {
-      {FileFormat.AVRO, 2},
-      {FileFormat.PARQUET, 2},
-      {FileFormat.ORC, 2}
+      {FileFormat.AVRO, 2, false},
+      {FileFormat.PARQUET, 2, true},
+      {FileFormat.PARQUET, 2, false},
+      {FileFormat.ORC, 2, true}
```
Comment (pvary) [inline flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java:79]: Why did you opt for this exact parametrization?

I happy for having tests for Parquet in both cases, as that is the main usecase for us now:
```
      {FileFormat.PARQUET, 2, true},
      {FileFormat.PARQUET, 2, false},
```

I might have opted for testing the future features, and keep one test for backward comp, like this:
```
      {FileFormat.AVRO, 2, true},
      {FileFormat.PARQUET, 2, true},
      {FileFormat.PARQUET, 2, false},
      {FileFormat.ORC, 2, true}
```

**Review by pvary — APPROVED** [review]
LGTM, some questions, where I have no strong opinion, but I wanted to raise them, so we decide on the consciously.

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterFileScanTaskReader.java (hunk 3 above)
Comment (stevenzwu) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/source/reader/ConverterFileScanTaskReader.java:29]: that is a great idea. will do

File: flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java (hunk 4 above)
Comment (stevenzwu) [reply flink/v1.20/flink/src/test/java/org/apache/iceberg/flink/source/TestIcebergSourceBoundedGenericRecord.java:79]: sounds good to me. will change
